flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -1727,6 +1764,160 @@ class JobManager(
+  }
+
+  private def instantiateClassLoaderMetrics(metrics: MetricGroup) {
+    val mxBean = ManagementFactory.getClassLoadingMXBean
+
+    metrics
+      .gauge("ClassesLoaded", new Gauge[Long] {
+        override def getValue: Long = mxBean.getTotalLoadedClassCount
+      })
+    metrics.gauge("ClassesUnloaded", new Gauge[Long] {
+      override def getValue: Long = mxBean.getUnloadedClassCount
+    })
+  }
+
+  private def instantiateGarbageCollectorMetrics(metrics: MetricGroup) {
+    val garbageCollectors = ManagementFactory.getGarbageCollectorMXBeans
+    
+    garbageCollectors.asScala foreach {
+      case gc =>
+        val gcGroup = metrics.addGroup("\"" + gc.getName + "\"")
+        gcGroup.gauge("Count", new Gauge[Long] {
+          override def getValue: Long = gc.getCollectionCount
+        })
+        gcGroup.gauge("Time", new Gauge[Long] {
+          override def getValue: Long = gc.getCollectionTime
+        })
+    }
+  }
+
+  private def instantiateMemoryMetrics(metrics: MetricGroup) {
+    val mxBean = ManagementFactory.getMemoryMXBean
+    val heap = metrics.addGroup("Heap")
+    heap.gauge("Used", new Gauge[Long] {
+      override def getValue: Long = mxBean.getHeapMemoryUsage.getUsed
+    })
+    heap.gauge("Committed", new Gauge[Long] {
+      override def getValue: Long = mxBean.getHeapMemoryUsage.getCommitted
+    })
+    heap.gauge("Max", new Gauge[Long] {
+      override def getValue: Long = mxBean.getHeapMemoryUsage.getMax
+    })
+
+    val nonHeap = metrics.addGroup("NonHeap")
+    nonHeap.gauge("Used", new Gauge[Long] {
+      override def getValue: Long = mxBean.getNonHeapMemoryUsage.getUsed
+    })
+    nonHeap.gauge("Committed", new Gauge[Long] {
+      override def getValue: Long = mxBean.getNonHeapMemoryUsage.getCommitted
+    })
+    nonHeap.gauge("Max", new Gauge[Long] {
+      override def getValue: Long = mxBean.getNonHeapMemoryUsage.getMax
+    })
+
+    val con = ManagementFactory.getPlatformMBeanServer;
+
+    val directObjectName = new ObjectName("java.nio:type=BufferPool,name=direct")
+
+    val direct = metrics.addGroup("Direct")
+    try {
+      direct.gauge("Count", new Gauge[Long] {
+        override def getValue: Long = con
+          .getAttribute(directObjectName, "Count").asInstanceOf[Long]
+      })
+      direct.gauge("MemoryUsed", new Gauge[Long] {
+        override def getValue: Long = con
+          .getAttribute(directObjectName, "MemoryUsed").asInstanceOf[Long]
+      })
+      direct.gauge("TotalCapacity", new Gauge[Long] {
+        override def getValue: Long = con
+          .getAttribute(directObjectName, "TotalCapacity").asInstanceOf[Long]
+      })
+    } catch {
+      case e: Exception =>
+      log.debug("Failed to add TaskManager metric.", e);
+    }
+
+    val mappedObjectName = new ObjectName("java.nio:type=BufferPool,name=mapped")
+
+    val mapped = metrics.addGroup("Mapped")
+    try {
+      mapped.gauge("Count", new Gauge[Long] {
```

Comment:
the number of buffers in the pool. the metric names were taken from https://docs.oracle.com/javase/7/docs/api/java/lang/management/BufferPoolMXBean.html